app/services/price_policies/time_based_price_calculator.rb

```diff
@@ -40,6 +45,47 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
       end
     end
 
+    def duration_pricing_cost_and_subsidy(duration_mins)
+      duration_in_hours = duration_mins / 60.0
+
+      result = build_intervals.reduce({ time_left: duration_in_hours, cost: 0, subsidy: 0 }) do |acc, interval_data|
+        time_left = acc[:time_left]
+
+        interval_length = interval_data[:interval_end] - interval_data[:interval_start]
+
+        time_to_charge = [time_left, interval_length].min
+
+        acc[:time_left] -= time_to_charge
+        acc[:cost] += interval_data[:rate] ? interval_data[:rate] * time_to_charge : 0
+        acc[:subsidy] += interval_data[:subsidy] ? interval_data[:subsidy] * time_to_charge : 0
```

Comment:
We should only ever have one or the other:
```suggestion
        if interval_data[:rate].present?
          acc[:cost] += interval_data[:rate] * time_to_charge : 0
        elsif interval_data[:subsidy].present?
          acc[:subsidy] +=  interval_data[:subsidy] * time_to_charge : 0
        end
```

Reply:
Based on what I see in `[cost_and_subsidy](https://github.com/tablexi/nucore-open/blob/master/app/services/price_policies/time_based_price_calculator.rb#L33-L41)`, we should set both